File: Data Structures & Algorithms/binary-tree-diameter/submission-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class MaxValue {
    int value = Integer.MIN_VALUE;
}

class Solution {

    public int diameterOfBinaryTree(TreeNode root) {
       MaxValue max = new MaxValue();
       diameterOfBinaryTreeH(root, max);
       return max.value;
    }

    public int diameterOfBinaryTreeH(TreeNode root, MaxValue max) {
        if(root == null) return 0;
        int left   = diameterOfBinaryTreeH(root.left, max);
        int right  = diameterOfBinaryTreeH(root.right,max);
        max.value = Math.max(max.value, left + right);
        return 1 + Math.max(left ,right);
    }
}

File: Data Structures & Algorithms/balanced-binary-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        self.isBalanced = True
        if root:
            self.height(root)
            return self.isBalanced
        else:
            return True    
    

    def height(self, root):
        if root:
            left = self.height(root.left)
            right = self.height(root.right)
            if abs(left - right) > 1: 
                self.isBalanced = False
            return 1 + max(left, right)    
        return 0    